src/algo/coloring/wfc.rs

```diff
@@ -0,0 +1,288 @@
+    NoValidConfiguration,
+}
+
+impl core::fmt::Display for WfcColoringError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            WfcColoringError::DirectedGraph => write!(f, "Graph must be undirected"),
+            WfcColoringError::NoValidConfiguration => {
+                write!(f, "No valid coloring configuration found")
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for WfcColoringError {}
+
+/// \[Generic\] [Wave Function Collapse algorithm][1] to properly color a non-weighted undirected graph.
+///
+/// This is a constraint satisfaction algorithm that assigns colors to vertices such that
+/// no adjacent vertices share the same color. The algorithm uses entropy-based heuristics
+/// to determine the order of vertex coloring and constraint propagation to ensure consistency.
+///
+/// The graph **must** be undirected. It should not contain loops.
+///
+/// # Arguments
+/// * `graph`: undirected graph without loops.
+///
+/// # Returns
+/// Returns a [`Result`] containing:
+/// * [`struct@std::collections::HashMap`] that associates to each `NodeId` its color (1-based numbering), or
+/// * [`WfcColoringError`]: error if the graph cannot be colored or is directed.
+///
+/// # Complexity
+/// * Time complexity: **O(|V|² × (|V| + |E|))** in worst case, **O(|V| × (|V| + |E|))** in best case.
+/// * Auxiliary space: **O(|V|²)**.
+///
+/// where **|V|** is the number of nodes and **|E|** is the number of edges.
+///
+/// [1]: https://arxiv.org/pdf/2108.09329
+///
+/// # Example
+/// ```rust
+/// use petgraph::{Graph, Undirected};
+/// use std::collections::HashMap;
+/// use petgraph::algo::wfc_coloring;
+///
+/// let mut graph = Graph::<(), (), Undirected>::new_undirected();
+/// let a = graph.add_node(());
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///     (a, b),
+///     (b, c),
+///     (c, a),
+/// ]);
+///
+/// // a ----- b
+/// // \      /
+/// //  \    /
+/// //   \  /
+/// //    c
+///
+/// let coloring = match wfc_coloring(&graph) {
+///     Ok(coloring) => coloring,
+///    Err(e) => panic!("Error: {}", e),
+/// };
+/// // [color-1] a ----- b [color-2]
+/// //           \      /
+/// //            \   /
+/// //              c [color-3]
+///
+/// assert_ne!(coloring[&a], coloring[&b]); // Adjacent vertices have different colors
+/// assert_ne!(coloring[&b], coloring[&c]); // Adjacent vertices have different colors
+/// assert_ne!(coloring[&c], coloring[&a]); // Adjacent vertices have different colors
+/// ```
+pub fn wfc_coloring<G>(graph: G) -> Result<HashMap<G::NodeId, usize>, WfcColoringError>
+where
+    G: IntoNeighbors + NodeCount + NodeIndexable + Visitable + GraphProp,
+    G::NodeId: Eq + Hash + Copy,
+{
+    if graph.is_directed() {
+        return Err(WfcColoringError::DirectedGraph);
+    }
+
+    let node_count = graph.node_count();
```

Comment:
```suggestion
    let node_count = graph.node_bound();
```
I just noticed this, but you can not assume that you can create a vertex with all vertices by using the node_count. Instead you need to use `node_bound`, since the indices used for the nodes might exceed node_count - 1 (in case of StableGraphs).